dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,209 @@
+  auto &insh = in_shapes[0];
+  ret.resize(insh.size() + 1);
+  for (int i = 0; i < axis; i++) {
+    ret[i] = insh[i];
+  }
+  ret[axis] = in_shapes.size();
+  for (int i = axis + 1; i < ret.size(); i++) {
+    ret[i] = insh[i - 1];
+  }
+  return ret;
+}
+
+
+template<>
+TensorShape<> DetermineShape<false>(span<const TensorShape<>> in_shapes, int axis) {
+  TensorShape<> ret = in_shapes[0];
+  ret[axis] = 0;
+  for (const auto &sh : in_shapes) {
+    ret[axis] += sh[axis];
+  }
+  return ret;
+}
+
+
+/**
+ * Joins multiple input tensors into one output tensor, along given axis.
+ *
+ * The kernel works in 2 modes: STACK and CONCAT. In CONCAT mode, kernel creates a new tensor,
+ * with joined values along given dimension, e.g.
+ *
+ * arr0 = [[1, 2, 4, 2], [1, 1, 7, 6], [6, 8, 8, 4]])
+ * shape = (3, 4)
+ *
+ * arr1 = [[3, 8, 8, 6], [8, 1, 5, 7], [6, 2, 7, 5]]
+ * shape = (3, 4)
+ *
+ * concatenate([arr0, arr1], axis=1) -> [[1, 2, 4, 2, 3, 8, 8, 6],
+ *                                       [1, 1, 7, 6, 8, 1, 5, 7],
+ *                                       [6, 8, 8, 4, 6, 2, 7, 5]]
+ * shape = (3, 8)
+ *
+ * Stacking, on the other hand, creates new tensor with added dimension, where `axis` points.
+ *
+ * stack([arr0, arr1], axis=1) -> [[[1, 2, 4, 2],
+ *                                  [3, 8, 8, 6]],
+ *
+ *                                 [[1, 1, 7, 6],
+ *                                  [8, 1, 5, 7]],
+ *
+ *                                 [[6, 8, 8, 4],
+ *                                  [6, 2, 7, 5]]]
+ * shape = (3, 2, 4)
+ *
+ * @tparam new_axis if true, STACK mode is applied.
+ */
+template<typename T, bool new_axis = true, int dims = -1>
+struct TensorJoinCpu {
+  ///@{
+  /**
+   * @param in_shapes Shapes of input tensors.
+   * @param axis Axis, along which tensors will be joined. Accepts positive and negative values
+   *             (e.g. axis: -1 <=> ndims-1)
+   */
+  KernelRequirements Setup(KernelContext &ctx, span<const TensorShape<dims>> in_shapes, int axis) {
+    n_input_tensors_ = in_shapes.size();
+    auto ndims = in_shapes[0].sample_dim();
+    DALI_ENFORCE(axis >= -ndims + !new_axis && axis <= ndims - !new_axis,
+                 make_string("Incorrect axis. Actual: ", axis, ". Expected in [",
+                             -ndims + !new_axis, ", ", ndims - !new_axis, "] interval (",
+                             new_axis ? "STACK" : "CONCAT", " mode)"));
+    axis_ = axis >= 0 ? axis : ndims + axis + new_axis;
+
+    {
+      const auto &ref = in_shapes[0];
+      for (int i = 1; i < n_input_tensors_; i++) {
+        DALI_ENFORCE(in_shapes[i].sample_dim() == ref.sample_dim(),
+                     "Every input shape must have the same number of dimensions.");
+        for (int j = 0; j < ref.size(); j++) {
+          if (!new_axis) {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j] || j == axis_, make_string(
+                    "Number of samples in every dimension (except the concatenated one) "
```

Comment:
```suggestion
                    "CONCAT: Number of samples in every dimension (except the concatenated one) "
```

Reply:
IMHO, the error would be clearer, if the mode info is at the end. Also, sole `CONCAT` or `STACK` I believe isn't that clear, it should be accompanied with `mode`, to remove any ambiguities. Given that, I'd prefer to remain with what I've wrote in the first place, if you don't mind?